src/cc/BPFTable.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <exception>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "bpf_module.h"
+
+namespace ebpf {
+
+class BPFTableBase {
+ public:
+  size_t max_size() {
+    return max_size_;
+  }
+
+ protected:
+  BPFTableBase(BPFModule* bpf_module, const std::string& name);
+
+  bool lookup(void* key, void* value);
+  bool next(void* key, void* next_key);
+
+  int fd_;
+  size_t max_size_;
+};
+
+template<class KeyType, class ValueType>
+class BPFHashTable : protected BPFTableBase {
+ public:
+  BPFHashTable(BPFModule* bpf_module, const std::string& name)
+    : BPFTableBase(bpf_module, name) {}
+
+  ValueType get_value(const KeyType& key) {
+    ValueType res;
+    if (!lookup(
+      const_cast<void*>(static_cast<const void*>(&key)),
+      static_cast<void*>(&res)
+    ))
+      throw std::invalid_argument("Key does not exist in the table");
+    return res;
+  }
+
+  ValueType operator[](const KeyType& key) {
+    return get_value(key);
+  }
+
+  std::vector<std::pair<KeyType, ValueType>> get_table_offline() {
+    std::vector<std::pair<KeyType, ValueType>> res;
+    res.reserve(max_size());
+
+    KeyType cur, nxt;
+    ValueType value;
+
+    while (true) {
+      if (!next(static_cast<void*>(&cur), static_cast<void*>(&nxt)))
+        break;
+      if (!lookup(static_cast<void*>(&nxt), static_cast<void*>(&value))) {
+        fprintf(stderr, "Failed reading table %d\n", fd_);
+        break;
+      }
+      res.push_back(std::make_pair(nxt, value));
+      std::swap(cur, nxt);
+    }
+
+    return res;
+  }
+};
+
+class BPFStackTable : protected BPFTableBase {
+ public:
+  BPFStackTable(BPFModule* bpf_module, const std::string& name)
+    : BPFTableBase(bpf_module, name) {}
+  ~BPFStackTable();
+
+  std::vector<uint64_t> get_stack_addr(int stack_id);
+  std::vector<std::string> get_stack_symbol(int stack_id, int pid);
+ private:
+  // From src/cc/export/helpers.h
```

Comment:
This is an installed header. If I reference it directly I would have to install that header as well, which would be confusing since that header is to be used **inside** BPF program code...